File: Asmi Bhaskar/day 20.c
#include <stdio.h>

int isPossible(int arr[], int n, int k, int mid) {
    int painters = 1;
    int currentLength = 0;

    for (int i = 0; i < n; ++i) {
        if (currentLength + arr[i] > mid) {
            painters++;
            currentLength = arr[i];
        } else {
            currentLength += arr[i];
        }
    }

    return (painters <= k);
}

int minTimeToPaint(int arr[], int n, int k) {
    int low = 0, high = 0;
    for (int i = 0; i < n; ++i) {
        high += arr[i];
        low = (arr[i] > low) ? arr[i] : low;
    }

    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (isPossible(arr, n, k, mid)) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    return low;
}

int main() {
    int arr[] = {10, 20, 30, 40};
    int n = sizeof(arr) / sizeof(arr[0]);
    int k = 2;

    int result = minTimeToPaint(arr, n, k);
    printf("Minimum time to paint all boards: %d\n", result);

    return 0;
}

File: Shailja/Day20.java
public class Day20 {
  public static void main(String[] args) {
            int[] arr = {10, 20, 30, 40};
            int n = arr.length;
            int k = 2;  
            System.out.println("Minimum time to paint all boards: " + findMinTime(arr, n, k));
        }
    
        public static int findMinTime(int[] arr, int n, int k) {
            int lo = getMax(arr, n);
            int hi = getSum(arr, n);
    
            while (lo < hi) {
                int mid = lo + (hi - lo) / 2;
                if (isFeasible(arr, n, k, mid)) {
                    hi = mid;
                } else {
                    lo = mid + 1;
                }
            }
    
            return lo;
        }
    
        public static boolean isFeasible(int[] arr, int n, int k, int maxTime) {
            int painters = 1;
            int currentTime = 0;
    
            for (int i = 0; i < n; i++) {
                currentTime += arr[i];
    
                if (currentTime > maxTime) {
                    painters++;
                    currentTime = arr[i];
    
                    if (painters > k) {
                        return false;
                    }
                }
            }
    
            return true;
        }
    
        public static int getMax(int[] arr, int n) {
            int max = arr[0];
            for (int i = 1; i < n; i++) {
                if (arr[i] > max) {
                    max = arr[i];
                }
            }
            return max;
        }
    
        public static int getSum(int[] arr, int n) {
            int sum = 0;
            for (int i = 0; i < n; i++) {
                sum += arr[i];
            }
            return sum;
        }
    }
    